### PR TITLE
Set environment wise API deployed time to API artifact

### PR DESCRIPTION
**Pull request by RAVEENSR**
This PR adds ability to set environment wise API deployed time to API artifact.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.wso2.carbon.apimgt.api.APIManagementException;
+import org.wso2.carbon.apimgt.impl.dao.GatewayArtifactsMgtDAO;
+import org.wso2.carbon.apimgt.impl.dto.APIRuntimeArtifactDto;
+import org.wso2.carbon.apimgt.impl.dto.OrgAndRevisionDeployedTimeInfoDTO;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Util {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java:28]:
```suggestion
public class ArtifactSynchronizerUtil {
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 2]
```diff
@@ -216,29 +217,39 @@ private void updateGatewayPublishedAPIArtifacts(Connection connection, String ap
...
         try (Connection connection = GatewayArtifactsMgtDBUtil.getArtifactSynchronizerConnection();
              PreparedStatement preparedStatement = connection.prepareStatement(query)) {
-            preparedStatement.setString(1, uuid);
+            preparedStatement.setString(1, apiUUID);
+            preparedStatement.setString(2, envName);
+            preparedStatement.setString(3, revisionUUId);
             try (ResultSet resultSet = preparedStatement.executeQuery()) {
                 while (resultSet.next()) {
                     organization = resultSet.getString("ORGANIZATION");
+                    deployedTime = resultSet.getString("DEPLOYED_TIME");
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:243]: Shall we use resultSet.getTimestamp method then no need to convert it to Date. We can use the long integer format date to the deployment.json file

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        String organization = orgAndDeployedTime.getOrganization();
+        String deployedTime = orgAndDeployedTime.getDeployedTime();
+        if (organization != null) {
+            apiRuntimeArtifactDto.setOrganization(orgAndDeployedTime.getOrganization());
+        }
+        if (deployedTime != null) {
+            try {
+                SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+                Date date = dateFormat.parse(deployedTime);
+                apiRuntimeArtifactDto.setDeployedTimeStamp(String.valueOf(date.getTime()));
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java:46]: I think we can use getTimestamp method in SQL level and avoid this.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.wso2.carbon.apimgt.impl.dto.OrgAndRevisionDeployedTimeInfoDTO;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Util {
+
+    private static final GatewayArtifactsMgtDAO gatewayArtifactsMgtDAO = GatewayArtifactsMgtDAO.getInstance();
+
+    public static void setOrgAndAPIRevisionDeployedTime(APIRuntimeArtifactDto apiRuntimeArtifactDto)
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java:32]: need to think of a better name than mentioning org id and revision time

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java:32]: Lets consider this on all other naming.

Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java:32]: Any suggestions for the method name? May be like `setAtrifactPropertyValues`, `setAtrifactProperties`, `setAdditionalAtrifactProperties`?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java (hunk 1 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java]: Fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java (hunk 2 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java (hunk 3 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java]: Fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java (hunk 4 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/Util.java]: Used `setArtifactProperties` as the method name.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 5]
```diff
@@ -2777,7 +2777,12 @@ public class SQLConstants {
             "DELETE FROM AM_GW_API_DEPLOYMENTS WHERE API_ID = ?";
     public static final String DELETE_GW_PUBLISHED_API_DETAILS = "DELETE FROM AM_GW_PUBLISHED_API_DETAILS WHERE " +
             "API_ID = ?";
-    public static final String RETRIEVE_ORGANIZATION = "SELECT ORGANIZATION FROM AM_API WHERE API_UUID =?";
+    public static final String RETRIEVE_ORGANIZATION_AND_API_REVISION_DEPLOYED_TIME =
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java:2780]: shall we fix this name too?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/ArtifactSynchronizerUtil.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    private static final GatewayArtifactsMgtDAO gatewayArtifactsMgtDAO = GatewayArtifactsMgtDAO.getInstance();
+
+    public static void setArtifactProperties(APIRuntimeArtifactDto apiRuntimeArtifactDto)
+            throws APIManagementException {
+        APIArtifactPropertyValues orgAndDeployedTime =
+                gatewayArtifactsMgtDAO.retrieveAPIArtifactPropertyValues(apiRuntimeArtifactDto.getApiId(),
+                        apiRuntimeArtifactDto.getLabel(), apiRuntimeArtifactDto.getRevision());
+        String organization = orgAndDeployedTime.getOrganization();
+        Timestamp deployedTime = orgAndDeployedTime.getDeployedTime();
+        if (organization != null) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/ArtifactSynchronizerUtil.java:38]: lets fix orgAndDeployedTime name too

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/dto/EnvironmentDto.java [hunk 7]
```diff
@@ -55,7 +66,7 @@ public boolean equals(Object obj) {
         }
 
         EnvironmentDto environmentDto = (EnvironmentDto) obj;
-        // check only name (environment name or label name), no need to consider vhost
+        // check only name (environment name or label name), no need to consider vhost and deployedTimeStamp
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/dto/EnvironmentDto.java:69]:
```suggestion
        // check only name (environment name or label name)
```

**Review by malinthaprasan — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java (hunk 5 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/ArtifactSynchronizerUtil.java (hunk 6 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/ArtifactSynchronizerUtil.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/dto/EnvironmentDto.java (hunk 7 above)
Comment (RAVEENSR) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/dto/EnvironmentDto.java]: fixed

**Pull request merged**